src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs

```diff
@@ -0,0 +1,157 @@
+            CborTag tag = PeekTagCore(out int additionalBytes);
+
+            AdvanceBuffer(1 + additionalBytes);
+            _isTagContext = true;
+            return tag;
+        }
+
+        public void ReadTag(CborTag expectedTag)
+        {
+            CborTag tag = PeekTagCore(out int additionalBytes);
+
+            if (expectedTag != tag)
+            {
+                throw new InvalidOperationException("CBOR tag mismatch.");
+            }
+
+            AdvanceBuffer(1 + additionalBytes);
+            _isTagContext = true;
+        }
+
+        public CborTag PeekTag() => PeekTagCore(out int _);
+
+        private CborTag PeekTagCore(out int additionalBytes)
+        {
+            CborInitialByte header = PeekInitialByte(expectedType: CborMajorType.Tag);
+            return (CborTag)ReadUnsignedInteger(_buffer.Span, header, out additionalBytes);
+        }
+
+        // Additional tagged type support
+
+        public DateTimeOffset ReadDateTimeOffset()
+        {
+            // implements https://tools.ietf.org/html/rfc7049#section-2.4.1
+
+            switch (PeekTag())
+            {
+                case CborTag.DateTimeString:
+                    ReadTag();
+
+                    if (Peek() != CborReaderState.TextString)
+                    {
+                        throw new FormatException("String DateTime semantic tag should annotate string value.");
+                    }
+
+                    string dateString = ReadTextString();
+
+                    if (!DateTimeOffset.TryParseExact(dateString, CborWriter.Rfc3339FormatString, null, DateTimeStyles.RoundtripKind, out DateTimeOffset result))
+                    {
+                        throw new FormatException("DateTime string is not valid RFC3339.");
+                    }
+
+                    return result;
+
+                case CborTag.DateTimeUnixSeconds:
+                    ReadTag();
+
+                    switch (Peek())
+                    {
+                        case CborReaderState.UnsignedInteger:
+                        case CborReaderState.NegativeInteger:
+                            return DateTimeOffset.FromUnixTimeSeconds(ReadInt64());
+
+                        case CborReaderState.HalfPrecisionFloat:
+                        case CborReaderState.SinglePrecisionFloat:
+                        case CborReaderState.DoublePrecisionFloat:
+                            // we don't (but probably should) have a float overload for DateTimeOffset.FromUnixTimeSeconds
+                            double seconds = ReadDouble();
+                            long epochTicks = DateTimeOffset.UnixEpoch.Ticks;
+                            long ticks = checked(epochTicks + (long)(seconds * TimeSpan.TicksPerSecond));
+                            return new DateTimeOffset(ticks, TimeSpan.Zero);
+
+                        default:
+                            throw new FormatException("Epoch DateTime semantic tag should annotate numeric value.");
```

Comment:
There's a loss of reader consistency on the exceptions after ReadTag()... the reader should probably rewrind.

FWIW, the fact that these methods can throw when Skip won't does reinforce for me the opinion that Skip shouldn't validate UTF-8 encodings for text string values.  (I think the real decision comes when conformance modes are added... does Skip verify that the integer `2` is not lazily encoded across multiple bytes, vs being in the compact single-byte form?  If yes, then Skip validating "primitive" values and not validating tagged values is consistent.  If no, then it should only validate what it needs to maintain state and the ability to jump to the next value).

Reply:
Agreed that adding conformance levels should add desired degree of control wrt Skip behaviour.